sdk/python/feast/go_server.py

```diff
@@ -249,96 +216,71 @@ def get_online_features(
-            signal.SIGTERM,
-            lambda sig, frame: self._go_server_background_thread.stop_go_server(),
+            signal.SIGTERM, lambda sig, frame: self._go_server_background_thread.stop(),
         )
         signal.signal(
-            signal.SIGINT,
-            lambda sig, frame: self._go_server_background_thread.stop_go_server(),
+            signal.SIGINT, lambda sig, frame: self._go_server_background_thread.stop(),
         )
 
         # Wait for go server subprocess to start for the first time before returning
         self._go_server_started.wait()
 
-    def _start_go_server(self):
-        if self._shared_connection.is_process_alive():
-            self._shared_connection.kill_process()
-
-        self._shared_connection.connect()
-        atexit.register(lambda: self._shared_connection.kill_process())
-        signal.signal(
-            signal.SIGTERM, lambda sig, frame: self._shared_connection.kill_process()
-        )
-        signal.signal(
-            signal.SIGINT, lambda sig, frame: self._shared_connection.kill_process()
-        )
-
     def kill_go_server_explicitly(self):
-        if self._check_use_thread():
-            self._go_server_background_thread.stop_go_server()
-        else:
-            self._shared_connection.kill_process()
+        self._go_server_background_thread.stop()
 
 
-# https://www.geeksforgeeks.org/python-different-ways-to-kill-a-thread/
-class GoServerBackgroundThread(threading.Thread):
+class GoServerMonitorThread(threading.Thread):
     def __init__(
         self,
         name: str,
         shared_connection: GoServerConnection,
-        go_server_started: threading.Event,
+        go_server_first_started: threading.Event,
     ):
         threading.Thread.__init__(self)
         self.name = name
         self._shared_connection = shared_connection
-        self._go_server_started = go_server_started
+        self._is_cancelled = False
+        self.daemon = True
+        self._go_server_started = go_server_first_started
 
     def run(self):
         # Target function of the thread class
+        _logger.info("Started GoServerMonitorThread")
         try:
-            while True:
-                self._go_server_started.clear()
+            while not self._is_cancelled:
 
                 # If we fail to connect to grpc stub, terminate subprocess and repeat
                 if not self._shared_connection.connect():
                     self._shared_connection.kill_process()
                     continue
-                self._go_server_started.set()
-                while True:
+                else:
+                    _logger.info("Set go server first started event")
+                    self._go_server_started.set()
+                while not self._is_cancelled:
                     try:
                         # Making a blocking wait by setting timeout to a very long time so we don't waste cpu cycle
                         self._shared_connection.wait_for_process(3600)
```

Comment:
So if we send a cancel command, how does that command kill the process. Isn't the loop blocked here?

Reply:
No, the cancel command will kill the process and the loop with break out because `wait_for_process` will return when the process ends. The cancel command will have set `is_cancelled` so the loop will terminate.